File: 第6次课/第6次课作业代码/第6次课作业代码/第一题 牛顿插值/1 下三角矩阵求逆/newtoninterpolation.py
import numpy as np
def newtoninterpolation(x,y,x1 ):
    #计算下三角矩阵A
    n=len(x);
    m=len(x1);
    A= np.zeros((n,n));
    #pi0为1
    A[:,0]=1;
    #递推计算其他
    for i in range(1,n):
        #下三角，所以计算一半加对角
        #第0行有了就不计算了
        #使用首项和最后一项检验项数
        for j in range(1,i+1):
            A[i,j]=A[i,j-1]*(x[i]-x[j-1])
   
    #计算系数矩阵a
    a=np.zeros((n,1));
    a[0]=y[0];
    #转换为矩阵，注意数据维度
    for i in range(1,n):
        a[i,0]=(y[i]- sum(A[i,0:i]*a[0:i,0]) )/A[i,i];
    a=np.matrix(a)
    #计算待求点系数矩阵A
    A2= np.zeros((m,n));
    A2[:,0]=1;
    #递推计算其他
    for i in range(0,m):#从0开始
        #这里不是下三角,注意范围
        for j in range(1,n-1):
            A2[i,j]=A2[i,j-1]*(x1[i]-x[j-1])
    #计算待求点值
    y1=(A2*a).T;
    y1=y1.tolist();
    return y1[0];

File: 第6次课/第6次课作业代码/第6次课作业代码/第一题 牛顿插值/1 下三角矩阵求逆/newtoninterpolation.m
function y = newtoninterpolation(x0,y0,x )

n=length(x0);
m=length(x);
A=zeros(n);    
A(:,1)=1;
for k=2:n
   j=2;
   A(k,j)=x0(k)-x0(1);
   for j=3:k
       A(k,j)=A(k,j-1)*(x0(k)-x0(j-1));
   end
end
a=invlowertriangular(A,y0')
% a=A\y0';

for l=1:m
    z=x(l);
    M=zeros(1,n);
    
    M(1)=1;
    for j=2:n
        M(j)=M(j-1)*(z-x0(j-1));
    end
%     y(l)=sum(M'.*a);
    y(l)=M*a;
end

end


